File: algorithms/1d_dynamic_programming/152-maximum_product_subarray.go
package main

import "fmt"

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

func maxProduct(nums []int) int {
	res := nums[0]
	curMin, curMax := 1, 1

	for i := 0; i < len(nums); i++ {
		tmp := curMax * nums[i] // Use old curMax
		curMax = max(nums[i], max(nums[i]*curMax, nums[i]*curMin))
		curMin = min(nums[i], min(nums[i]*tmp, nums[i]*curMin))
		res = max(res, curMax)
	}

	return res
}

func main() {
	nums := []int{2, 3, 0, -2, 4, -1}
	// nums := []int{-2, 0, -1}
	fmt.Println(maxProduct(nums))
}

File: algorithms/python-algorithms/1d_dynamic_programming/152-maximum_product_subarray.py
from typing import List


# 2023-05-04
# TC: O(n) - Iterate through nums array only once
# SC: O(1) - Storing max integers only, no extra ds needed
class Solution:
    def maxProduct(self, nums: List[int]) -> int:
        res = nums[0]
        curMin, curMax = 1, 1

        for n in nums:
            tmp = curMax * n  # Use old curMax

            curMax = max(n, tmp, n * curMin)
            curMin = min(n, tmp, n * curMin)

            res = max(res, curMax)

        return res


def main():
    pass


if __name__ == "__main__":
    main()
